src/ffi2/datetime.rs

```diff
@@ -0,0 +1,354 @@
+        microsecond: c_int,
+        tzinfo: *mut PyObject,
+        cls: *mut PyTypeObject,
+    ) -> *mut PyObject,
+    pub Delta_FromDelta: unsafe extern "C" fn(
+        days: c_int,
+        seconds: c_int,
+        microseconds: c_int,
+        normalize: c_int,
+        cls: *mut PyTypeObject,
+    ) -> *mut PyObject,
+    pub DateTime_FromTimestamp:
+        unsafe extern "C" fn(cls: *mut PyTypeObject, args: *mut PyObject, kwargs: *mut PyObject)
+            -> *mut PyObject,
+    pub Date_FromTimestamp:
+        unsafe extern "C" fn(cls: *mut PyTypeObject, args: *mut PyObject) -> *mut PyObject,
+}
+
+// Type struct wrappers
+
+const _PyDateTime_DATE_DATASIZE: usize = 4;
+const _PyDateTime_TIME_DATASIZE: usize = 6;
+const _PyDateTime_DATETIME_DATASIZE: usize = 10;
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct PyDateTime_Date {
+    pub ob_base: PyObject,
+    pub hashcode: Py_hash_t,
+    pub hastzinfo: c_char,
+    pub data: [c_uchar; _PyDateTime_DATE_DATASIZE],
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct PyDateTime_Time {
+    pub ob_base: PyObject,
+    pub hashcode: Py_hash_t,
+    pub hastzinfo: c_char,
+    pub data: [c_uchar; _PyDateTime_TIME_DATASIZE],
+    pub tzinfo: *mut PyObject,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct PyDateTime_DateTime {
+    pub ob_base: PyObject,
+    pub hashcode: Py_hash_t,
+    pub hastzinfo: c_char,
+    pub data: [c_uchar; _PyDateTime_DATETIME_DATASIZE],
+    pub tzinfo: *mut PyObject,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct PyDateTime_Delta {
+    pub ob_base: PyObject,
+    pub hashcode: Py_hash_t,
+    pub days: c_int,
+    pub seconds: c_int,
+    pub microseconds: c_int,
+}
+
+// C API Capsule
+// Note: This is "roll-your-own" lazy-static implementation is necessary because
+// of the interaction between the call_once locks and the GIL. It turns out that
+// calling PyCapsule_Import releases and re-acquires the GIL during the import,
+// so if you have two threads attempting to use the PyDateTimeAPI singleton
+// under the GIL, it causes a deadlock; what happens is:
+//
+// Thread 1 acquires GIL
+// Thread 1 acquires the call_once lock
+// Thread 1 calls PyCapsule_Import, thus releasing the GIL
+// Thread 2 acquires the GIL
+// Thread 2 blocks waiting for the call_once lock
+// Thread 1 blocks waiting for the GIL
+//
+// However, Python's import mechanism acquires a module-specific lock around
+// each import call, so all call importing datetime will return the same
+// object, making the call_once lock superfluous. As such, we can weaken
+// the guarantees of the cache, such that PyDateTime_IMPORT can be called
+// until __PY_DATETIME_API_UNSAFE_CACHE is populated, which will happen exactly
+// one time. So long as PyDateTime_IMPORT has no side effects (it should not),
+// this will be at most a slight waste of resources.
+static __PY_DATETIME_API_ONCE: Once = Once::new();
+static mut __PY_DATETIME_API_UNSAFE_CACHE: *const PyDateTime_CAPI = ptr::null();
```

Comment:
On the rust side, you don't need the two underscores. Unless you declare an item or struct field pub, it's invisible outside the module